### PR TITLE
Add git shell API keys

### PR DESCRIPTION
**Pull request by eugenk**
This is the backend-part of https://github.com/ontohub/ontohub-models/pull/160.
It integrates the system-tests of https://github.com/ontohub/system-test/pull/43, too, which in turn use the git-shell at https://github.com/ontohub/git-shell/pull/2.

# Review Notes

I suggest to review the pull requests in this order:
1. https://github.com/ontohub/ontohub-models/pull/160
1. https://github.com/ontohub/ontohub-backend/pull/389
1. https://github.com/ontohub/git-shell/pull/2
1. https://github.com/ontohub/system-test/pull/43

And keep in mind my comment in https://github.com/ontohub/git-shell/pull/2 about how the git-shell works in general. Also look at the git-shell specific tests in the travis-run of the system-tests to see the use cases.

### CODE REVIEW

**Comment by eugenk** [comment]
There is an error with the hets-agent in the system-tests. This is because we have not installed Hets on Travis. As soon as there is an up-to-date Ubuntu package, we will install it there.

**Comment by phyrog** [comment]
Heh when you think you're finished with a review, but it was only the first commit :roll_eyes:

File: config/initializers/core_extensions/devise/strategies/api_key.rb [hunk 1]
```diff
@@ -7,7 +7,10 @@ module Strategies
     # Strategy for API key authentication
     class ApiKey < HttpAuthorization
       def authenticate!
-        api_key = ::ApiKey.
+        api_key = ::HetsApiKey.
```
[the remark this answers is not in the document]
Comment (eugenk) [reply config/initializers/core_extensions/devise/strategies/api_key.rb]: This is needed: They have different authentications because of different salts.

**Comment by eugenk** [comment]
I think it's best if you review all changes at once because I stumbled a few times and needed to change old changes again...

File: app/policies/git_shell_policy.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+# Policies for the GitShell
+class GitShellPolicy < ApplicationPolicy
+  def authorize?
+    current_user.is_a?(GitShellApiKey)
```
Comment (phyrog) [inline app/policies/git_shell_policy.rb]: `git_shell_api_key?`

File: spec/policies/organizational_unit_policy_spec.rb [hunk 3]
```diff
@@ -35,11 +35,26 @@
...
-    it 'does allow show?' do
-      expect(subject.show?).to be(true)
+    context 'HetsApiKey' do
+      let(:current_user) { create(:hets_api_key) }
+      subject do
+        OrganizationalUnitPolicy.new(current_user)
+      end
+
+      it 'does allow show?' do
+        expect(subject.show?).to be(true)
```
[the remark this answers is not in the document]
Comment (eugenk) [reply spec/policies/organizational_unit_policy_spec.rb]: Correct. I changed it a few commits later.

[the remark this answers is not in the document]
Comment (eugenk) [reply spec/policies/organizational_unit_policy_spec.rb]: That's why it's better to not review commit by commit.

File: spec/policies/repository_policy_spec.rb [hunk 4]
```diff
@@ -537,35 +537,52 @@
...
+
+      %i(create? update? index? destroy? write?).each do |method|
+        it "does not allow #{method}" do
+          expect(subject.public_send(method)).to be(false)
         end
       end
 
-      it 'returns all repositories' do
-        expect(subject.resolve.to_a.size).to eq(5)
+      it "does #{show_expectation || 'not '}allow show?" do
```
[the remark this answers is not in the document]
Comment (eugenk) [reply spec/policies/repository_policy_spec.rb]: Fixed with bb8eaa1c

File: lib/refs_updater.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    end
+  end
+
+  protected
+
+  def analyze(_ref, before, after)
+    # This assumes that it was _not_ a force-push. It can _only_ be done because
+    # force-pushes are prevented. With a force-push, we would need to find the
+    # first common ancestor C of `before` and `after`, analyze all commits from
+    # C onwards and delete `FileVersion`s after C leading to `before`.
```
Comment (phyrog) [inline lib/refs_updater.rb]: Unless these commits are also part of another branch

Comment (eugenk) [reply lib/refs_updater.rb]: Fixed with 2c2de71f

File: app/policies/git_shell_policy.rb (hunk 2 above)
Comment (eugenk) [reply app/policies/git_shell_policy.rb]: Fixed with 2c2de71f

File: lib/refs_updater.rb (hunk 5 above)
Comment (phyrog) [reply lib/refs_updater.rb]: Sorry, I meant that FileVersions for commits that are also part of another branch should not be deleted then.

Comment (eugenk) [reply lib/refs_updater.rb]: Fixed with 923c50b2. Note that I added a new commit (87ba412d) right before that one.

**Comment by eugenk** [comment]
```
/home/travis/system-test/repositories/ontohub-backend/app/models/authorized_keys_file.rb:5:in `<class:AuthorizedKeysFile>': undefined method `join' for "/home/travis/system-test/tmp/data":String (NoMethodError)
```

Why is it a String and not a Pathname?

**Comment by phyrog** [comment]
Not this again...

**Comment by phyrog** [comment]
I think it just needs to be added to https://github.com/ontohub/ontohub-backend/blob/923c50b299607ca0bc6e12a9bff984f67b185cca/config/initializers/settings_normalizer.rb#L18

**Comment by eugenk** [comment]
But it is already there:
https://github.com/ontohub/ontohub-backend/blob/923c50b299607ca0bc6e12a9bff984f67b185cca/config/initializers/settings_normalizer.rb#L18-L19

See also
https://github.com/ontohub/ontohub-backend/blob/923c50b299607ca0bc6e12a9bff984f67b185cca/app/models/authorized_keys_file.rb#L4-L5

**Comment by eugenk** [comment]
Maybe the file is loaded in the production environment before the settings are normalized.

**Comment by eugenk** [comment]
d02627db should fix it, if that's really the problem.

**Comment by eugenk** [comment]
It fixed the error, but there is a new one: Travis cannot SSH-connect to localhost. The authorized_keys file seems to be broken. I will debug this after the lunch break

**Comment by eugenk** [comment]
The system-tests are green. They use https://github.com/ontohub/system-test/pull/46.

**Comment by eugenk** [comment]
See ae07a70.

**Pull request merged**